### ISSUE TITLE
GBA games delay starting after BIOS when movie is active

### ISSUE THREAD

**Issue by EZGames69** [TASVideos/BizHawk#1479]
I noticed it took a few seconds for the game to start after the bios if a movie is being played, but when a movie isnt being played, it does not have this issue.  One game that I see this happening on is Donkey Kong Country 2.  

Video with no movie active: https://www.youtube.com/watch?v=f3BQrE_pTT4

Video with a movie active: https://www.youtube.com/watch?v=msTSlixj808

Video of how it starts up on a DS (if you need me to show it starting on a real GBA I can do that, I just used DS because it's backlit): https://www.youtube.com/watch?v=4m5fOSk9gbo

I havent tested if this happens in other games yet, but I do believe it happens to some of the super mario advanced games like smw and simmilar.

EDIT: also I am not sure if this has been reported before, so sorry if I'm bringing up old issues.

**Comment by YoshiRulz** [comment]
Which GBA core are you using?

**Comment by EZGames69** [comment]
mGBA

**Comment by vadosnaprimer** [comment]
Try in actual mGBA.

**Comment by EZGames69** [comment]
on mGBA 0.7, the issue is not present.

the delay only happens when a movie is being played back or if TAStudio is open.

**Comment by vadosnaprimer** [comment]
Good. Please check mGBA 0.6. If it's there, it will be fixed when we update.

**Comment by EZGames69** [comment]
any specific one? like 0.6.3 or somthing?

**Comment by EZGames69** [comment]
I wasnt able to get the bios to load on any 0.6 version.

**Comment by synnae** [comment]
> 
> 
> any specific one? like 0.6.3 or somthing?

BizHawk 2.3.1 uses mGBA 0.6.3, according to the emulator's About window:
https://imgur.com/L78AjZc

I also managed to load the bios on said mGBA version. To guarantee it will work, make sure both the ROM and bios are No Intro dumps, then in the emulator, go "Tools > Settings > BIOS" and put the file there.

The delay problem doesn't happen there either btw, so it seems to be a BizHawk exclusive issue. Here is a video if anyone needs:
[mGBA 0.6.3.zip](https://github.com/TASVideos/BizHawk/files/2833944/mGBA.0.6.3.zip)

**Comment by EZGames69** [comment]
yep, it's the exact same as on 0.7

I believe it is a bizhawk issue only because it happens when playing a movie.

**Comment by MittenzHugg** [comment]
I run into the same issue with Ty the Tasmanian Tiger 2 when playing a movie/using TAStudio

**Comment by Asnivor** [comment]
This might be something to do with the convoluted way MGBAHawk handles DeterministicEmulation (latest dev build):

https://github.com/TASVideos/BizHawk/blob/master/BizHawk.Emulation.Cores/Consoles/Nintendo/GBA/MGBAHawk.cs#L17

A value of `deterministicEmulation == true` is only passed into the core constructor if a movie is envolved in any way. Note: this is actually true of all cores that don't specify and handle a DeterministicEmulation settings as a SyncSetting.

Looking at the MGBAHawk core constructor, determinism will be enforced if:

- A movie is involved in any way
- A movie is not involved *but* a BIOS *is* present, `RTCUseRealTime == false` and `SkipBios == false`

At all other times deterministicemulation will be false. The only 2 things this affects in MBGA appear to be `skipbios` and `RTCUseRealTime`.

SkipBios *will* happen if the user has this set in syncsettings and a movie isnt involved. Not sure that this has anything to do with what we are seeing in this issue though (all those videos show a bios right?):

https://github.com/TASVideos/BizHawk/blob/master/BizHawk.Emulation.Cores/Consoles/Nintendo/GBA/MGBAHawk.cs#L38

RTCUseRealTime *will* happen if a user has this set in syncsettings and again, a movie isnt involved:

https://github.com/TASVideos/BizHawk/blob/master/BizHawk.Emulation.Cores/Consoles/Nintendo/GBA/MGBAHawk.cs#L208

Now, I don't actually know how this realtimeclock setting would affect MGBA, but I guess it could introduce the lack of delay we are seeing?

Someone who can actually test this, can you try changing the syncsettings to:

- RTCUseRealTime = false
- SkipBios = false

and running the game *without a movie* to see if this then looks the same as with a movie? This should inadvertently force the core into `DeterministicEmulation == true` (if I am reading the code right).

TLDR: you are probably always going to see differences between movie and non-movie in this core because it doesnt manage a SyncSettings option for DeterministicEmulation (unless your other syncsettings are 'just right').

This probably needs to be addressed in some way regardless of whether this is the actual issue or not...

**Comment by Asnivor** [comment]
Actually, can you also try in the latest dev build (if you havent already)? There was a deterministic fix/reversion in feb2019.

https://ci.appveyor.com/project/zeromus/bizhawk-udexo/build/artifacts

**Comment by EZGames69** [comment]
I just tried it in that build and the issue still applies

**Comment by Asnivor** [comment]
Did you try with `RTCUseRealTime = false` and `SkipBios = false` ?

**Comment by EZGames69** [comment]
just did that, and still no change.  (I'm testing with DKC2 GBA for refrence).

**Comment by Asnivor** [comment]
hmm, I dunno then.

**Comment by EZGames69** [comment]
Tested in the newest dev build with the newest mGBA and the bios delay is still in it.

**Comment by vadosnaprimer** [comment]
I don't understand this at all. If I set Real Time RTC and BIOS Skip to false, this forces deterministic mode, but without tastudio or movie running, it still doesn't have additional frames there. Sync mode doesn't affect this. Skip BIOS is the only setting sent to the core in that regard, and again, without movie running it doesn't cause a delay. But it's indeed frames mGBA is emulating. Moreover, it's different across games. In _Donkey Kong Country 2_ it's ~120 frames, in _DK - King of Swing_ - only 4... I tried returning OS time while movie is running, still got extra frames.

**Comment by EZGames69** [comment]
thanks for at least trying.  yeah it's very strange, at this point I dont really expect it to be fixed anytime soon.

**Comment by ThunderAxe31** [comment]
This is **not** a bug. Donkey Kong Country 2 takes more time when you record a movie because it starts from _clean SRAM_. Try deleting the SRAM file or doing right click > "Close and Clear SRAM", then load the ROM again and you will see it taking the extra frames just like when recording a movie, regardless of the BIOS skip setting. This is because most GBA games take their time for initializing SRAM data when they detect that it's corrupted or blank (0xFF bytes padded). Case closed.
